src/kernel/pgtbl.c

```diff
@@ -4,184 +4,145 @@
+{
+	return chal_cap_memactivate(ct, pt, frame_cap, dest_pt, vaddr);
+}
 
-	/* pa2va (value in *kern_addr) will return NULL if the page is
-	 * not kernel accessible */
-	if (unlikely(!*kern_addr)) return -EINVAL; /* cannot retype a non-kernel accessible page */
-	if (unlikely(retypetbl_kern_ref((void *)(new_v & PGTBL_FRAME_MASK)))) return -EFAULT;
+int
+pgtbl_activate(struct captbl *t, unsigned long cap, unsigned long capin, pgtbl_t pgtbl, u32_t lvl)
+{
+	return chal_pgtbl_activate(t, cap, capin, pgtbl, lvl);
+}
 
-	/* We keep the cos_frame entry, but mark it as COSKMEM so that
-	 * we won't use it for other kernel objects. */
-	if (unlikely(cos_cas((unsigned long *)pte, orig_v, new_v) != CAS_SUCCESS)) {
-		/* restore the ref cnt. */
-		retypetbl_deref((void *)(orig_v & PGTBL_FRAME_MASK));
-		return -ECASFAIL;
-	}
-	/* Return the pte ptr, so that we can release the page if the
-	 * kobj activation failed later. */
-	*pte_ret = (unsigned long *)pte;
+int
+pgtbl_deactivate(struct captbl *t, struct cap_captbl *dest_ct_cap, unsigned long capin,
+                 livenessid_t lid, capid_t pgtbl_cap, capid_t cosframe_addr, const int root)
+{
+	return chal_pgtbl_deactivate(t, dest_ct_cap, capin, lid, pgtbl_cap, cosframe_addr, root);
+}
 
-	return 0;
+/**
+ * this works on both kmem and regular user memory: the retypetbl_ref
+ * works on both.
+ */
+int
+pgtbl_mapping_add(pgtbl_t pt, u32_t addr, u32_t page, u32_t flags)
+{
+	return chal_pgtbl_mapping_add(pt, addr, page, flags);
 }
 
-/* Return 1 if quiescent past since input timestamp. 0 if not. */
+/**
+ * This function is only used by the booting code to add cos frames to
+ * the pgtbl. It ignores the retype tbl (as we are adding untyped
+ * frames). 
+ */
 int
-tlb_quiescence_check(u64_t timestamp)
+pgtbl_cosframe_add(pgtbl_t pt, u32_t addr, u32_t page, u32_t flags)
 {
-	int i, quiescent = 1;
-
-	/* Did timer interrupt (which does tlb flush
-	 * periodically) happen after unmap? The periodic
-	 * flush happens on all cpus, thus only need to check
-	 * the time stamp of the current core for that case
-	 * (assuming consistent time stamp counters). */
-	if (timestamp > tlb_quiescence[get_cpuid()].last_periodic_flush) {
-		/* If no periodic flush done yet, did the
-		 * mandatory flush happen on all cores? */
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			if (timestamp > tlb_quiescence[i].last_mandatory_flush) {
-				/* no go */
-				quiescent = 0;
-				break;
-			}
-		}
-	}
-	if (quiescent == 0) {
-		printk("from cpu %d, t %llu: cpu %d last mandatory flush: %llu\n", get_cpuid(), timestamp, i,
-		       tlb_quiescence[i].last_mandatory_flush);
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			printk("cpu %d: flush %llu\n", i, tlb_quiescence[i].last_mandatory_flush);
-		}
-	}
-
-	return quiescent;
+	return chal_pgtbl_cosframe_add(pt, addr, page, flags);
 }
 
+/* This function updates flags of an existing mapping. */
```

Comment:
These comments are spot-on.  They're great:  they provide context that cannot really be found from the arguments or name.

In the future, I think that we still want types that are specific to flags and what-not for better self-documentation, but we can wait for that.